jfr/src/main/java/io/netty/jfr/JfrInstrumentedByteBufAllocator.java

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright 2020 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.jfr;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.CompositeByteBuf;
+import io.netty.util.internal.ObjectUtil;
+import jdk.jfr.Category;
+import jdk.jfr.Description;
+import jdk.jfr.Enabled;
+import jdk.jfr.Event;
+import jdk.jfr.Label;
+import jdk.jfr.Name;
+
+import static io.netty.jfr.JfrInstrumentedByteBufAllocator.AllocationType.*;
+
+
+/**
+ * A {@link ByteBufAllocator} that creates JDK Flight Recorder events for all byte buffer allocations.
+ */
+public final class JfrInstrumentedByteBufAllocator implements ByteBufAllocator {
+
+    private final ByteBufAllocator wrapped;
+
+    private JfrInstrumentedByteBufAllocator(ByteBufAllocator wrapped) {
+        this.wrapped = ObjectUtil.checkNotNull(wrapped, "wrapped");
+    }
+
+    public static ByteBufAllocator wrap(ByteBufAllocator wrapped) {
+        return new JfrInstrumentedByteBufAllocator(wrapped);
+    }
+
+    @Override
+    public ByteBuf buffer() {
+        ByteBufAllocationEvent event = new ByteBufAllocationEvent();
```

Comment:
Same question as above, as buffer operation are likely to be many. Can these new event allocation be eliminated, maybe using the same trick above with the `ThreadLocal` ?